File: frontend/src/lib/shared/__tests__/base.test.ts
import { describe, expect, it, vi } from 'vitest';
import fs from 'fs';
import path from 'path';

describe('Shared Logic Infrastructure', () => {
  describe('Directory Structure', () => {
    it('has normalizers directory', () => {
      const normalizersPath = path.join(__dirname, '../normalizers');
      expect(fs.existsSync(normalizersPath)).toBe(true);
    });

    it('has validators directory', () => {
      const validatorsPath = path.join(__dirname, '../validators');
      expect(fs.existsSync(validatorsPath)).toBe(true);
    });

    it('has mutators directory', () => {
      const mutatorsPath = path.join(__dirname, '../mutators');
      expect(fs.existsSync(mutatorsPath)).toBe(true);
    });

    it('has services directory', () => {
      const servicesPath = path.join(__dirname, '../services');
      expect(fs.existsSync(servicesPath)).toBe(true);
    });
  });

  describe('Module Exports', () => {
    it('exports from normalizers index', async () => {
      const normalizers = await import('../normalizers');
      expect(normalizers).toBeDefined();
    });

    it('exports from validators index', async () => {
      const validators = await import('../validators');
      expect(validators).toBeDefined();
    });

    it('exports from mutators index', async () => {
      const mutators = await import('../mutators');
      expect(mutators).toBeDefined();
    });
  });

  describe('Base Classes', () => {
    describe('BaseNormalizer', () => {
      it('defines normalize interface', async () => {
        const { BaseNormalizer } = await import('../normalizers/base-normalizer');
        const normalizer = new BaseNormalizer();
        expect(normalizer.normalize).toBeDefined();
      });

      it('throws error for base class normalize', async () => {
        const { BaseNormalizer } = await import('../normalizers/base-normalizer');
        const normalizer = new BaseNormalizer();
        expect(() => normalizer.normalize({})).toThrow('Must be implemented by subclass');
      });
    });

    describe('BaseValidator', () => {
      it('defines validate interface', async () => {
        const { BaseValidator } = await import('../validators/base-validator');
        const validator = new BaseValidator();
        expect(validator.validate).toBeDefined();
      });

      it('returns validation result structure', async () => {
        const { BaseValidator } = await import('../validators/base-validator');
        
        class TestValidator extends BaseValidator {
          validate(value: any, context?: any): { valid: boolean; errors?: Record<string, string[]> } {
            return { valid: true };
          }
        }

        const validator = new TestValidator();
        const result = validator.validate('test');
        expect(result).toHaveProperty('valid');
        expect(result.valid).toBe(true);
      });
    });

    describe('BaseMutator', () => {
      it('defines mutate interface', async () => {
        const { BaseMutator } = await import('../mutators/base-mutator');
        const mutator = new BaseMutator();
        expect(mutator.mutate).toBeDefined();
      });

      it('passes through data by default', async () => {
        const { BaseMutator } = await import('../mutators/base-mutator');
        
        class TestMutator extends BaseMutator {
          mutate(data: any, context?: any): any {
            return data;
          }
        }

        const mutator = new TestMutator();
        const data = { name: 'test' };
        expect(mutator.mutate(data)).toEqual(data);
      });
    });
  });
});

File: test/lib/shared/base_test.rb
require "test_helper"

class SharedLogicInfrastructureTest < ActiveSupport::TestCase
  test "normalizers directory exists" do
    assert File.directory?(Rails.root.join("lib/shared/normalizers"))
  end

  test "validators directory exists" do
    assert File.directory?(Rails.root.join("lib/shared/validators"))
  end

  test "mutators directory exists" do
    assert File.directory?(Rails.root.join("lib/shared/mutators"))
  end

  test "loads BaseNormalizer" do
    assert_nothing_raised { Shared::Normalizers::BaseNormalizer }
  end

  test "loads BaseValidator" do
    assert_nothing_raised { Shared::Validators::BaseValidator }
  end

  test "loads BaseMutator" do
    assert_nothing_raised { Shared::Mutators::BaseMutator }
  end

  test "BaseNormalizer defines normalize interface" do
    assert_respond_to Shared::Normalizers::BaseNormalizer, :normalize
  end

  test "BaseNormalizer raises NotImplementedError for base class" do
    assert_raises(NotImplementedError) do
      Shared::Normalizers::BaseNormalizer.normalize("test")
    end
  end

  test "BaseValidator defines validate interface" do
    assert_respond_to Shared::Validators::BaseValidator, :validate
  end

  test "BaseValidator returns validation result structure" do
    # Create a concrete validator for testing
    test_validator = Class.new(Shared::Validators::BaseValidator) do
      def self.validate(value, context = {})
        validation_result(valid: true)
      end
    end

    result = test_validator.validate("test")
    assert_includes result, :valid
    assert_includes result, :errors
    assert result[:valid]
    assert_equal({}, result[:errors])
  end

  test "BaseMutator defines mutate interface" do
    assert_respond_to Shared::Mutators::BaseMutator, :mutate
  end

  test "BaseMutator raises NotImplementedError for base class" do
    assert_raises(NotImplementedError) do
      Shared::Mutators::BaseMutator.mutate({})
    end
  end

  test "BaseMutator passes through data in concrete implementation" do
    # Create a concrete mutator for testing
    test_mutator = Class.new(Shared::Mutators::BaseMutator) do
      def self.mutate(data, context = {})
        data
      end
    end

    data = { name: "test" }
    assert_equal data, test_mutator.mutate(data)
  end
end
